File: Sean的leetcode/Leetcode/Binary Tree Right Side View.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#if 1 // DFS
/* 1. traverse (right subtree first) root->right->left.
   2. target_level is the output level.
   3. curr_level is the level of the root.
 */
class Solution {
public:
    void dfs(TreeNode *root, vector<int> &res, int level)
    {
        if (!root) return;

        if (level == res.size())
            res.push_back(root->val);

        dfs(root->right, res, level+1);
        dfs(root->left, res, level+1);
    }

    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        dfs(root, res, 0);
        return res;
    }
};
#else //BFS
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        queue<pair<int, TreeNode*> >bfs;
        vector<int>res;
        pair<int, TreeNode*>node;

        if (!root) return res;

        bfs.push(make_pair(0, root));

        while (!bfs.empty())
        {
            node = bfs.front();

            if (node.second->left)
                bfs.push(make_pair(node.first+1, node.second->left));

            if (node.second->right)
                bfs.push(make_pair(node.first+1, node.second->right));

            bfs.pop();

            if (!bfs.empty() && bfs.front().first != node.first)
                res.push_back(node.second->val);
        }
        res.push_back(node.second->val);
        return res;
    }
};
#endif

File: MyLeetCode/python/Binary Tree Right Side View.py
###
# Time Complexity: O(n)
# Space Complexity: O(n)
###
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        q = collections.deque([root])
        result = []
        while q:
            level_cnt = len(q)
            for i in xrange(level_cnt):
                cur = q.popleft()
                if i == level_cnt - 1:
                    result.append(cur.val)
                if cur.left:
                    q.append(cur.left)
                if cur.right:
                    q.append(cur.right)
        return result
